crates/pico_core/src/dependency_stack.rs

```diff
@@ -0,0 +1,30 @@
+use std::cell::RefCell;
+
+#[derive(Debug, Default)]
+pub struct DependencyStack<T>(RefCell<Vec<Vec<T>>>);
+
+impl<T> DependencyStack<T> {
+    pub fn new() -> Self {
+        Self(RefCell::new(Vec::new()))
+    }
+
+    pub fn enter(&self) {
```

Comment:
nit: perhaps we return a guard that, when dropped, pops the item off the stack?

Reply:
Oh, that's neat! We can protect a case when we forgot to call `leave`